enhancements/machine-config/os-images-streams.md

```diff
@@ -0,0 +1,1075 @@
+> is fully functional.
+
+**Backward Compatibility:**
+
+The old single-stream format continues to work:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: machine-config-osimageurl
+  namespace: openshift-machine-config-operator
+data:
+  baseOSContainerImage: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  baseOSExtensionsContainerImage: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  releaseVersion: "4.21.0"
+```
+
+When `streams.json` is not present, the MCO automatically creates a default stream
+with a version-specific name based on the release (e.g., `"rhel9-coreos"` for OpenShift 4.x)
+using the `baseOSContainerImage` and `baseOSExtensionsContainerImage` fields. This ensures
+existing clusters continue to function without modification.
+
+##### Stream Merging and Default Stream Selection
+
+The MCO implements a `StreamSource` interface that allows composition and merging
+of streams from multiple sources. The merging process:
+
+1. **Collection**: Each source is queried for its available streams
+2. **Merging**: Streams are merged into a map keyed by stream name. When multiple
+sources provide the same stream name, the later source (higher precedence) wins
+3. **Default Stream Identification**: The MCO searches for a stream with a
+version-specific hardcoded name based on the distribution and release version:
+   - RHCOS in OpenShift 4.x: `"rhel9-coreos"`
+   - RHCOS in OpenShift 5.0+: `"rhel10-coreos"`
+   - FCOS: `"fedora-coreos"` (unversioned for Fedora's rolling release model)
+   - SCOS: `"stream-coreos"`
+4. **Validation**: If the default stream is not found in the merged streams, an
+error is returned
+
+**Important**: The default stream name is **version-specific and hardcoded** based on the
+MCO's build target and release version, not read from the ConfigMap's `"default"` field in
+`streams.json`. The ConfigMap's `"default"` field is informational only. This ensures
+explicit version selection and prevents silent OS version changes during platform upgrades.
+
+The implementation ensures:
+- Best-effort processing: if a source fails to provide streams, it's skipped with
+logging
+- Conflict logging: when streams are overridden, the conflict is logged for debugging
+- Handling of partial stream data (OS image without Extensions or vice versa)
+
+Implementation is located in `pkg/controller/osimagestream/osimagestream.go` and
+`BuildOSImageStreamFromSources()`.
+
+##### Default Stream Evolution and Upgrade Behavior
+
+The default stream name changes between OpenShift releases to reflect the recommended
+OS version for new clusters, while ensuring existing clusters maintain their current OS
+version during platform upgrades.
+
+**OpenShift 4.x Releases (4.21-4.23):**
+- Default stream: `"rhel9-coreos"`
+- Available streams: `"rhel9-coreos"`, `"rhel10-coreos"` (Tech Preview)
+- New clusters install with RHCOS 9
+- Existing clusters remain on their current stream
+
+**OpenShift 5.0+ Releases:**
```

Comment:
again, let's not talk about specific openshift versions, since that has not been decided anywhere